cli/flox/src/commands/services/mod.rs

```diff
@@ -71,35 +99,105 @@ impl ServicesCommands {
     }
 }
 
-/// Return a ConcreteEnvironment for variants that support services.
-pub fn supported_concrete_environment(
-    flox: &Flox,
-    environment: &EnvironmentSelect,
-) -> Result<ConcreteEnvironment> {
-    let concrete_environment = environment.detect_concrete_environment(flox, "Services in")?;
-    if let ConcreteEnvironment::Remote(_) = concrete_environment {
-        return Err(ServicesCommandsError::RemoteEnvsNotSupported.into());
+/// An augmented [ConcreteEnvironment] that has been checked for services support.
+/// Constructing a [ServicesEnvironment] requires a _local_ [ConcreteEnvironment]
+/// that supports services, i.e. is defined in a v1 [TypedManifestCatalog].
+///
+/// Constructing a [ServicesEnvironment] for a remote environment is not supported.
+///
+/// The [ServicesEnvironment] provides methods to guard
+pub struct ServicesEnvironment {
+    environment: ConcreteEnvironment,
+    socket: PathBuf,
+    manifest: TypedManifestCatalog,
+}
```

Comment:
I think it's alright since it consolidates `supported_environment`, `supported_concrete_environment`, the new guards, and a guaranteed way to get the socket path in one place.

Side thoughts: I've found "make a new type and impls" for _everything_ a bit jarring, but I'm also not super keen on how many separate utility functions we have in the services provider. Maybe just grouping them in separate modules would be sufficient.

Reply:
The newtype pattern is a pretty standard pattern in rust to establish and maintain invariants though the typesystem.
With bare `ConcreteEnvironment`s its fully up to the command impl to validate the concrete env type, read, parse and validate the manifest.
We already reparse and validate manifests defensively: [1]
The comment there hint at an implicit invariant "Checks for supported environments within the commands should prevent us ever getting here", but we cant really trust it anyway so we test again "just in case".
Using a newtype that provides this guarantee once its been parsed, prevents this entirely.

The guard methods express somewhat different guarantees over the state of the process rather than the environment itself.
I implemented on `ServicesEnvironment` as they do take data from it, but I can see them as module functions.


[1] https://github.com/flox/flox/blob/a850aa1db4d46fd82462dcf73d234a816aac60ae/cli/flox/src/commands/services/mod.rs#L254-L258

Reply:
I don't think the extra complexity is warranted. I think there's a lot of overlap between a `ServicesEnvironment` and a `PathEnvironment`. Services are a feature of an environment, not a core abstraction like "where the environment lives" or "the common functionality for all environments". I also think trying to guarantee much of anything about services is fraught with peril given that all kinds of things can change _while the command is executing_.

Reply:
Just to put a finer point on this, we already have
- UninitializedEnvironment
- CoreEnvironment
- ConcreteEnvironment
- PathEnvironment
- ManagedEnvironment
- RemoteEnvironment

I don't think that making a 7th (!!) environment type is making the situation better, and I don't know that wrappers are how we want to compose these things together. Furthermore I'm very wary of treating the guards that are here as invariants like you normally would with guards.

When you lock a mutex and get a guard back, you have _very_ high confidence that the resource protected by the mutex is exclusively yours. I don't think we have the same kind of guarantees from the guards presented here.

Presumably you would call `guard_processes_running` or similar, get the guard, and expect that afterwards the processes would be running, but since we don't control that or even get notifications when this changes, you don't really have knowledge of whether this guard will be valid immediately after you call the guard function. This means that any abstractions you build on top of having the guard is very iffy. So, if you can't rely on the guard for invariants, it's unclear that there's much to be gained from building abstractions on top of them since you'll either:
- Still need to check for correctness
- Assume correctness and get into a bad state when an "invariant" is violated.

Reply:
`Path-`, `Managed-` and `RemoteEnvironments` really only describe the different behaviors of "a flox environment" in different localities.
I'd argue since Managed environments now have a local backing as well, Path and Managed could be merged eventually if we were concerned about the duplicate effort there, but that's a straying thought..

the three environments implement _shared_ functionality, i.e. modifying manifest files through transactions, evaling, building, linking the environment store path,... through a _shared_ abstraction concerning **only** the `manifest.*` files.
You can discuss whether lumping shared data for all these methods into `CoreEnvironment` or spreading that out into functions makes a meaningful difference, but composing with `CoreEnvironment` seems to mostly work well.

`ConcreteEnvironment` is just bundling all three environment types together into a single type since some environments implement methods outside of the shared trait.
I personally dont think inflating the shared trait to a union of functionality and making it object safe could really remove the need for `ConcreteEnvironment` as is.

`UninitializedEnvironment` isn't really an environments at all its more a description of _where_ an environment is, and it only exists because we cant refer to remote environments by a local path (or at least not without requiring runtime checks).

The fact that all end with `Environment` is unfortunate and may be misleading as it suggests there are 6 different environments at the same level, while there are clear boundaries between them:

```mermaid
flowchart TD
    
uninit["UninitializedEnvironment"]
concrete["ConcreteEnvironment"]

path["PathEnvironment"]
managed["ManagedEnvironment"]
remote["RemoteEnvironment"]

core["CoreEnvironment"]
    
uninit -- "dereferenced into" --> concrete

concrete -. "either" .-> path
concrete -. "either" .-> managed
concrete -. "either" .-> remote

path -- shared functionality --> core
managed -- shared functionality --> core
remote  -- shared functionality --> core
```

`ServicesEnvironment` in this case is not a _new_ environment either but describes the requirements to the specific environment that can support services, i.e. 1) has to be local and 2) has to be v1.
"Services are a feature of an environment", yes, a specific subset of environments
Having _a_ type that describes and guarantees this seems safer than passing around concrete environments and doing that check repeatedly in different places or forgetting about it and depending on implicit invariants.
I'd agree that the argument for this type is much weaker if all environment localities support services (or both local environments are merged into a single concrete type that supports envs) and support for v0 manifests is dropped.


-----

Wrt to the `guard_processes_running` methods, i commented below that apparently process-compose does have specific error messaging for unresponsive/dead sockets and we can use that instead of adding a guard method.
That's to say i fully agree with you assessment there.

The guard method was the simpler solution to provide some guarantees and avoid checking for socket liveness with _every_ process compose interaction.
I think a lot of discussion around this resulted from misunderstanding process-composes behavior and how to work around this within flox.